### ISSUE TITLE
modelling of the dss SourceBus 

### ISSUE THREAD

**Issue by sanderclaeys** [lanl-ansi/PowerModelsDistribution.jl#88]
First of all, I would like to thank you all for developing this package; this can potentially be very helpful for my research.

I have a question regarding the modelling of the SourceBus, as defined in the dss format. My understanding of the current code is the following:

The SourceBus is set as the slack bus, with as voltage reference the specified value in p.u. The internal impedance of the SourceBus (specified through mvasc1 and mvasc3) only affects the ubound and lbound for the voltage magnitude of the SourceBus.

When running tp_pf, this leads to different results than OpenDSS. A while ago I developed a Python tool to parse dss files and calculate the unbalanced PF. I only reached the same results as OpenDSS after doing the following:

Replace the SourceBus by an additional line, of which the impedance is set according to the values of mvasc1 and mvasc3/ isc1 and isc3. Bus1 of this new line becomes the SourceBus, and bus2 of this new line replaces the old SourceBus.

I see in the code that the parameters of this new line are calculated, but not used for anything besides setting the ubound and lbound for the voltage magnitude of the SourceBus. The data files you use have very stiff sources, so this difference would not be noticable in the results for these cases.

**Is there a reason why the SourceBus is modelled as such and not by adding a line explicitly?** I created a new version which does exactly this, but the pf becomes infeasible when I increase the power demand, and I have not been able to figure out why so far.

**Comment by ccoffrin** [comment]
The short answer to this question is that we are still figuring out the best ways to map OpenDSS data into our much simpler internal data model.  Thus far, the mvasc1, mvasc3,  isc1, and isc3 parameters have not been a high priority.

@frederikgeth, if you agree this is the best interoperation of these, I will consider updating our parser.

@pseudocubic, any thoughts?

**Comment by ccoffrin** [comment]
I prefer option 1 for the implementation.

My more pressing questions for the moment are,
1. Is this feature a high priority in the near future? Or can we work around not having it?
2. If we do support it, what should the default interoperation be for the OpenDSS parser?  To match OpenDSS or be more numerically stable.

**Comment by frederikgeth** [comment]
- I'm not a big fan of replicating this voltage-source-with-internal-impedance just because OpenDSS does it, as I think it's mainly a work-around for the algorithm they use. As the algorithms have become better, we should consider not defaulting to such shortcuts anymore. This is further complicated due to the fact that people tend to conflate modelling artefacts which serve as workarounds for algorithms (e.g. this voltage source with internal impedance) with models of physical systems/components (e.g. generators, interface to a higher voltage level grid). 
- An OpenDSS compatibility mode is useful for validation, but replicating all OpenDSS functionality is not an immediate objective of this project. (although we welcome contributions in this direction)
- Whether or not compatibility mode is a default will depend on where TPPMs OpenDSS parser is going to position itself w.r.t extensions. E.g. if you can't easily extend an OpenDSS case with new parameters for optimisation models (e.g. cost factors of battery use and wear to minimize a household electricity bill with PV and storage), then the main use of the parser is to provide compatibility w.r.t. OpenDSS. In that case the default may be to replicate OpenDSS and give a warning that this may be numerically challenging. (This problem is challenging in the same way that modelling breakers in matpower format through lines with very small impedances sometimes leads to issues).

**Comment by hakanergun** [comment]
In principle each voltage source has an internal impedance in some way (generators, batteries, higher voltage grids), so I am not against having an internal impedance of the voltage source. As long as we do not explicitly model  internal generator voltages and the synchronous reactance, we will need an impedance to be physically correct. I am not sure if it would be possible to do both, ideal voltage sources and voltage sources with an impedance?

**Comment by frederikgeth** [comment]
By the way, @sanderclaeys, if the purpose is to validate TPPMs calculation w.r.t OpenDSS, you can try to over-write the parsed voltage bounds at the source-bus with the value calculated by OpenDSS. You can then call the run_tp_pf and should then more similar or even identical results (assuming the angles between the phases don't change much from 120 degrees due to the internal impedance). E.g.:
```
using ThreePhasePowerModels
using PowerModels
using Ipopt
data = ThreePhasePowerModels.parse_file("./test/data/opendss/case3_unbalanced.dss")
data["bus"]["1"]["vmin"] = [1 1 1] 
data["bus"]["1"]["vmax"] = [1 1 1]
run_tp_pf(data, ACPPowerModel, IpoptSolver())
```
(latest quick guide see https://lanl-ansi.github.io/ThreePhasePowerModels.jl/latest/quickguide.html)
Does this suit you?

What about this approach:
- @sanderclaeys makes the call whether this feature is a priority given his use-case (it isn't for me, but it isn't too hard to do, and simplifies validation). 
- If it is implemented, make it the default for now, but write out a warning that this may lead to issues, and display how to set an setting to disable this translation?

**Comment by sanderclaeys** [comment]
Apologies for the delay, I first wanted to see whether I could produce results consistent with OpenDSS by including the suggested approach. This I managed finally yesterday.

@frederikgeth for me this feature is not high priority, since I implemented a workaround myself. Also, my use case has a transformer just after the source, so in any case I need further customizations to achieve the same results as OpenDSS. However, I think there should be at least a warning indicating the current behaviour, that the internal impedance of the source is not modelled. 

@ccoffrin I could implement the proposed approach in a fork, and activate this behaviour internally in the parser by an optional keyword argument that defaults to false. Is that consistent with the philosophy of the package? 

The reason that the power flow did not converge in my case, is that the dss parser silently assumes that the units of the line and linecode are the same, which is not the case in the input file that I used. Hence, my lines were 1000x ("m" specified instead of "km" for lines) longer than I thought, rendering the pf infeasible.
